File: Soundbox.Reloaded.Ui.Web/scripts/services/playbackViewService.ts
module soundbox.services {
    import SoundDto = soundbox.domain.sounds.SoundDto;
    import PlayableSoundDto = soundbox.domain.sounds.PlayableSoundDto;

    export interface IPlaybackViewService {
        getQueue(): ng.IHttpPromise<SoundDto[]>;
        getCurrentSound(): ng.IHttpPromise<PlayableSoundDto>;
        isLooping(): ng.IHttpPromise<boolean>;
        setLooping(loop: boolean): ng.IHttpPromise<{}>;
        playSoundFromQueue(): ng.IPromise<{}>;
        playPause(): ng.IPromise<{}>;
        seek(seconds: number): ng.IPromise<{}>;
        addToQueue(sound: SoundDto): ng.IPromise<{}>;
        removeFromQueue(index: number): ng.IPromise<{}>;
    }

    export class PlaybackViewService implements IPlaybackViewService {
        private http: ng.IHttpService;

        constructor($http: ng.IHttpService) {
            this.http = $http;
        }

        public getQueue() {
            return this.http.get<SoundDto[]>('/api/sound/getQueue');
        }

        public getCurrentSound() {
            return this.http.get<PlayableSoundDto>('/api/sound/getCurrentSound');
        }

        public isLooping() {
            return this.http.get<boolean>('/api/sound/isLooping');
        }

        public setLooping(loop: boolean) {
            return this.http.post('/api/sound/setLooping?loop=' + loop, {});
        }

        public playSoundFromQueue() {
            return this.http.post('/api/sound/playSoundFromQueue', {});
        }

        public playPause() {
            return this.http.post('/api/sound/playPause', {});
        }

        public seek(seconds: number) {
            return this.http.post('/api/sound/seek?seconds=' + seconds, {});
        }

        public addToQueue(sound: SoundDto) {
            return this.http.post('/api/sound/addToQueue?soundId=' + sound.Id, {});
        }

        public removeFromQueue(index: number) {
            return this.http.post('/api/sound/removeFromQueue?index=' + index, {});
        }
    }

    angular.module('soundbox.services.playback', [])
        .service('playbackViewService', PlaybackViewService);
}

File: Soundbox.Reloaded.Ui.Presentation/ViewServices/PlaybackViewService.cs
namespace Soundbox.Reloaded.Ui.Presentation.ViewServices
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    using AutoMapper;

    using Soundbox.Reloaded.Core.Domain.Sounds.Model;
    using Soundbox.Reloaded.Core.Domain.Sounds.Objects;
    using Soundbox.Reloaded.Infrastructure.Services;
    using Soundbox.Reloaded.Infrastructure.DbAccess;
    using Soundbox.Reloaded.Ui.Presentation.Dto.Sounds;

    public class PlaybackViewService : IDisposable
    {
        private readonly UnitOfWork unitOfWork;
        private static bool soundEndsEventSubscribed = false;

        public PlaybackViewService(Action soundEndsAction)
        {
            this.unitOfWork = new UnitOfWork();

            if (soundEndsEventSubscribed == false)
            {
                QueuePlaybackService.Instance().SoundPlaybackService.SoundEnds += new SoundEndsEventHandler((object sender, EventArgs e) => soundEndsAction());
                soundEndsEventSubscribed = true;
            }
        }

        public IEnumerable<SoundDto> GetQueue()
        {
            return QueuePlaybackService.Instance().GetQueue()
                .Select(Mapper.Map<Sound, SoundDto>);
        }

        public PlayableSoundDto GetCurrentSound()
        {
            var sound = QueuePlaybackService.Instance().SoundPlaybackService.CurrentSoundOrDefault();
            return Mapper.Map<PlayableSound, PlayableSoundDto>(sound);
        }

        public bool IsLooping()
        {
            return QueuePlaybackService.Instance().IsLooping();
        }

        public void SetLooping(bool loop)
        {
            QueuePlaybackService.Instance().SetLooping(loop);
        }

        public void PlaySoundFromQueue()
        {
            QueuePlaybackService.Instance().PlaySoundFromQueue();
        }
        
        public bool PlayPause()
        {
            if (QueuePlaybackService.Instance().SoundPlaybackService.IsPlaying)
            {
                QueuePlaybackService.Instance().SoundPlaybackService.PauseSound();
                return false;
            }
            else
            {
                QueuePlaybackService.Instance().SoundPlaybackService.ContinueSound();
                return true;
            }
        }

        public void Seek(int seconds)
        {
            QueuePlaybackService.Instance().SoundPlaybackService.SeekSound(seconds);
        }

        public void AddToQueue(Guid soundId)
        {
            var sound = this.unitOfWork.SoundRepository.GetById(soundId);
            QueuePlaybackService.Instance().AddSound(sound);
        }

        public void RemoveFromQueue(int index)
        {
            QueuePlaybackService.Instance().RemoveSoundByIndex(index);
        }

        public void Dispose()
        {
            this.unitOfWork.Dispose();
        }
    }
}
